### PR TITLE
Add new input validator API functions

### LINKED ISSUE #1549: Assertion 'lit_is_cesu8_string_valid (string_p, string_size)' failed  in ecma_new_ecma_string_from_utf8
###### Jerry version:

```
Checked revision: 7b59c80
Build: debug.linux
```
###### OS:

```
Ubuntu 16.04, x86_64
```
###### Test case:
[UPDATE]
The test case contains special characters that GH cannot display, but it can be saved with this small python snippet (enough to paste it to a py interpreter):

```python
with open('test.js', 'wb') as f:
    f.write(b"function $() {\
$ = { \
$:$, $: ($ ($, ($ * ($[(($ + '(Cq\xd4c=V\xcc'.$ ())) & $] + $))))\
}\
}")
```
[ORIGINAL]
```javascript

function $() {

$ = {
$:$, $: ($ ($, ($ * ($[(($ + '(Cqc=V'.$ ())) & $] + $))))
}
}
```

###### Backtrace:

```
ICE: Assertion 'lit_is_cesu8_string_valid (string_p, string_size)' failed at jerryscript/jerry-core/ecma/base/ecma-helpers-string.c(ecma_new_ecma_string_from_utf8):154.
Error: ERR_FAILED_INTERNAL_ASSERTION
Aborted
bt
#0  0x000000000046daba in syscall_2 () at jerryscript/jerry-libc/target/posix/jerry-asm.S:59
#1  0x0000000000402760 in raise (sig=6) at jerryscript/jerry-libc/target/posix/jerry-libc-target.c:91
#2  0x0000000000402732 in abort () at jerryscript/jerry-libc/target/posix/jerry-libc-target.c:77
#3  0x00000000004113c6 in jerry_port_fatal (code=ERR_FAILED_INTERNAL_ASSERTION) at jerryscript/targets/default/jerry-port-default-fatal.c:53
#4  0x000000000042ce6f in jerry_fatal (code=ERR_FAILED_INTERNAL_ASSERTION) at jerryscript/jerry-core/jrt/jrt-fatals.c:61
#5  0x000000000042cec1 in jerry_assert_fail (assertion=0x47c788 "lit_is_cesu8_string_valid (string_p, string_size)", file=0x47c700 "jerryscript/jerry-core/ecma/base/ecma-helpers-string.c", function=0x480e30 <__func__.3095.lto_priv.558> "ecma_new_ecma_string_from_utf8", line=154) at jerryscript/jerry-core/jrt/jrt-fatals.c:85
#6  0x000000000045ae6c in ecma_new_ecma_string_from_utf8 (string_p=0x712be0 <buffer+2496> "B!\262E\264' + '(Cq\324c=V\314'.indexOf(strvar5))) & 255] + ary[((((true instanceof ((typeof EvalError == 'function') ? EvalError : Object)) >= 0 ? (true instanceof ((typeof EvalError == 'function') ? EvalError :"..., string_size=5) at jerryscript/jerry-core/ecma/base/ecma-helpers-string.c:154
#7  0x0000000000413790 in ecma_find_or_create_literal_string (chars_p=0x712be0 <buffer+2496> "B!\262E\264' + '(Cq\324c=V\314'.indexOf(strvar5))) & 255] + ary[((((true instanceof ((typeof EvalError == 'function') ? EvalError : Object)) >= 0 ? (true instanceof ((typeof EvalError == 'function') ? EvalError :"..., size=5) at jerryscript/jerry-core/ecma/base/ecma-literal-storage.c:72
#8  0x000000000042a4c6 in parser_post_processing (context_p=0x7fffffffcbc0) at jerryscript/jerry-core/parser/js/js-parser.c:1715
#9  0x000000000042b3c8 in parser_parse_function (context_p=0x7fffffffcbc0, status_flags=6) at jerryscript/jerry-core/parser/js/js-parser.c:2155
#10 0x0000000000425380 in lexer_construct_function_object (context_p=0x7fffffffcbc0, extra_status_flags=6) at jerryscript/jerry-core/parser/js/js-lexer.c:1639
#11 0x000000000041e347 in parser_parse_function_statement (context_p=0x7fffffffcbc0) at jerryscript/jerry-core/parser/js/js-parser-statm.c:396
#12 0x0000000000420d32 in parser_parse_statements (context_p=0x7fffffffcbc0) at jerryscript/jerry-core/parser/js/js-parser-statm.c:1688
#13 0x000000000042a99c in parser_parse_source.lto_priv.202 (source_p=0x712220 <buffer> "//", '-' <repeats 103 times>, "\n// Copyright (C) Microsoft. All rights reserved.\n// Licensed under the MIT license. See LICENS"..., size=2915, strict_mode=0, error_location_p=0x7fffffffcd70) at jerryscript/jerry-core/parser/js/js-parser.c:1874
#14 0x00000000004227b1 in parser_parse_script (source_p=0x712220 <buffer> "//", '-' <repeats 103 times>, "\n// Copyright (C) Microsoft. All rights reserved.\n// Licensed under the MIT license. See LICENS"..., size=2915, is_strict=false, bytecode_data_p=0x7fffffffcdb0) at jerryscript/jerry-core/parser/js/js-parser.c:2252
#15 0x00000000004699fe in jerry_parse (source_p=0x712220 <buffer> "//", '-' <repeats 103 times>, "\n// Copyright (C) Microsoft. All rights reserved.\n// Licensed under the MIT license. See LICENS"..., source_size=2915, is_strict=false) at jerryscript/jerry-core/jerry.c:262
#16 0x000000000046d580 in main (argc=3, argv=0x7fffffffd2e8) at jerryscript/jerry-main/main-unix.c:706
```

### PR DESCRIPTION
**Pull request by LaszloLango**
Fixes #1549

JerryScript-DCO-1.0-Signed-off-by: László Langó llango.u-szeged@partner.samsung.com

### CODE REVIEW

**Comment by robertsipka** [comment]
LGTM (informally).

File: jerry-core/jerry.c [hunk 1]
```diff
@@ -256,6 +256,11 @@ jerry_parse (const jerry_char_t *source_p, /**< script source */
 #ifdef JERRY_JS_PARSER
   jerry_assert_api_available ();
 
+  if (!lit_is_utf8_string_valid ((lit_utf8_byte_t *) source_p, (lit_utf8_size_t) source_size))
+  {
+    return ecma_raise_syntax_error (ECMA_ERR_MSG ("The input should be a valid UTF8 string."));
```
Comment (zherczeg) [inline jerry-core/jerry.c]: Input must be a valid UTF-8 string.

File: jerry-core/jerry-snapshot.c [hunk 2]
```diff
@@ -458,6 +458,11 @@ jerry_parse_and_save_snapshot (const jerry_char_t *source_p, /**< script source
                                size_t buffer_size) /**< the buffer's size */
 {
 #ifdef JERRY_ENABLE_SNAPSHOT_SAVE
+  if (!lit_is_utf8_string_valid ((lit_utf8_byte_t *) source_p, (lit_utf8_size_t) source_size))
+  {
+    return 0;
+  }
+
```
Comment (zherczeg) [inline jerry-core/jerry-snapshot.c]: Do we need this? jerry_parse_and_save_snapshot calls jerry_parse.

Comment (LaszloLango) [inline jerry-core/jerry-snapshot.c]: Yes we do. It does not call `jerry_parse`. It calls `parser_parse_script`

**Comment by zherczeg** [comment]
I have been thinking about this in the weekend. Perhaps we should make this a runtime enable.disable feature, and do check for all functions which accepts utf8/cesu8 input. This patch is more like a workaround than a fix.

**Comment by LaszloLango** [comment]
@zherczeg, @robertsipka I've updated the PR. Please review again.

File: docs/02.API-REFERENCE.md [hunk 3]
```diff
@@ -1548,9 +1561,12 @@ jerry_substring_to_utf8_char_buffer (const jerry_value_t value,
 
 **See also**
 
-- [jerry_create_string_from_utf8](#jerry_create_string)
+- [jerry_create_string_from_utf8](#jerry_create_string_from_utf8)
 - [jerry_get_utf8_string_size](#jerry_get_utf8_string_size)
 - [jerry_get_utf8_string_length](#jerry_get_utf8_string_length)
+- [jerry_is_utf8_string_valid](#jerry_is_utf8_string_valid)
+
+
```
Comment (zherczeg) [inline docs/02.API-REFERENCE.md]: Do we need two newlines here?

File: docs/02.API-REFERENCE.md [hunk 4]
```diff
@@ -3287,6 +3309,100 @@ bool foreach_function (const jerry_value_t prop_name,
...
+**Summary**
+
+Validate UTF-8 string.
+
+**Prototype**
+
+```c
+bool
+jerry_is_utf8_string_valid (const jerry_char_t *utf8_buf_p, /**< UTF-8 string */
```
Comment (zherczeg) [inline docs/02.API-REFERENCE.md]: What do you think of jerry_is_valid_utf8_string? Somehow that is more natural to me.

File: jerry-core/jerry-snapshot.c (hunk 2 above)
Comment (zherczeg) [inline jerry-core/jerry-snapshot.c]: Perhaps this is really not needed anymore.

File: docs/02.API-REFERENCE.md [hunk 5]
```diff
@@ -3287,6 +3309,100 @@ bool foreach_function (const jerry_value_t prop_name,
...
+    jerry_release_value (string_value);
+  }
+
+  jerry_cleanup ();
+}
+```
+
+**See also**
+
+- [jerry_run_simple](#jerry_run_simple)
```
Comment (zherczeg) [inline docs/02.API-REFERENCE.md]: jerry_run_simple expects utf8 string, not cesu8

File: docs/02.API-REFERENCE.md (hunk 3 above)
Comment (LaszloLango) [inline docs/02.API-REFERENCE.md]: Yes, we do. This is the style in the whole MD

File: docs/02.API-REFERENCE.md (hunk 4 above)
Comment (LaszloLango) [inline docs/02.API-REFERENCE.md]: Good to me. Should we rename the internal functions too? `lit_is_valid_utf8_string`?

Comment (zherczeg) [inline docs/02.API-REFERENCE.md]: At some point we should. If you wish you can do it in this patch.

Comment (LaszloLango) [inline docs/02.API-REFERENCE.md]: Ok, I'll do it in this patch.

File: docs/02.API-REFERENCE.md (hunk 5 above)
Comment (LaszloLango) [inline docs/02.API-REFERENCE.md]: Fixed, removed the reference from `jerry_is_cesu8_string_valid`

File: jerry-core/jerry.c [hunk 6]
```diff
@@ -2007,6 +2007,34 @@ jerry_foreach_object_property (const jerry_value_t obj_val, /**< object value */
 } /* jerry_foreach_object_property */
 
 /**
+ * Validate UTF-8 string
+ *
+ * @return true - if UTF-8 string is well-formed
+ *         false - otherwise
+ */
+bool
+jerry_is_utf8_string_valid (const jerry_char_t *utf8_buf_p, /**< UTF-8 string */
```
Comment (zherczeg) [inline jerry-core/jerry.c]: Please rename this function as well (jerry_is_valid_utf8_string)

File: jerry-core/jerryscript.h [hunk 7]
```diff
@@ -330,6 +330,12 @@ bool jerry_foreach_object_property (const jerry_value_t obj_val, jerry_object_pr
                                     void *user_data_p);
 
 /**
+ * Input validatator functions
+ */
+bool jerry_is_utf8_string_valid (const jerry_char_t *utf8_buf_p, jerry_size_t buf_size);
```
Comment (zherczeg) [inline jerry-core/jerryscript.h]: Here as well.

File: docs/02.API-REFERENCE.md [hunk 8]
```diff
@@ -3287,6 +3309,99 @@ bool foreach_function (const jerry_value_t prop_name,
 - [jerry_object_property_foreach_t](#jerry_object_property_foreach_t)
 
 
+# Input validatator functions
```
Comment (bzsolt) [inline docs/02.API-REFERENCE.md]: `validator`

File: jerry-core/jerryscript.h [hunk 9]
```diff
@@ -330,6 +330,12 @@ bool jerry_foreach_object_property (const jerry_value_t obj_val, jerry_object_pr
                                     void *user_data_p);
 
 /**
+ * Input validatator functions
```
Comment (bzsolt) [inline jerry-core/jerryscript.h]: `validator`

**Pull request merged**